### PR TITLE
Added Firebase Successfully with all desired functionalities

### PR DESCRIPTION
**Pull request by lakshyajn**
Hey @Satyam1923, Thank U for assigning me this project :)

#15 was my issue
### I have successfully added the Firebase Database in the project.
*Please give me highest score 🥇 

**I have created 3 files firebase.js, database.js and auth.js for calling the firebase ref, collection of data and authentication of the users respectively, and made changes in index.js accordingly,
Also added firebase files in package-lock.json and package.json files.**

**Here are some screenshots as well->**
Screenshot 2024-05-11 011143

So, I have made a firebase named "Spring", included my api key, ids,  domain and database urls for testing.
I tried and tested it and it is working fine.

**Firebase->**

DataBase Url: "https://Spring-default-rtdb-us-central1.firebasedatabase.app"
We can add any other url and credentials to the code thus making it modifiable easily.
If firebase crashes, we can switch to another!!

Screenshot 2024-05-11 011826

So,  I have added->
User Registration, User Sign in, User Sign Out, Error Handling for all three
Then I have added functionality to add, update, show, remove data from database,
and most importantly user Uthentication to check sign in, sign out

Thus our goals of firebase->
- Data collection
- User authentication
- User Registration
- User Profile Management are now done!!

Now the remaining part is just UI, from where user would give his/her email and password and that would be stored in backend and then database....

### CODE REVIEW

**Comment by Satyam1923** [comment]
Could you tell me how you have tested. So I can also take a look into it.

**Comment by sudoerp** [comment]
Please refer to the "Creating Pull Request" Section of the README.md before opening another PR.

**Comment by lakshyajn** [comment]
> Could you tell me how you have tested. So I can also take a look into it.

@Satyam1923 I first added storage bucket as u said.
That would be working to store the data, i.e., data collection
But that will not help in testing, coz we need to upload files of data in it first.

So I used realtime database, to see the results
I wrote the functions calls with required arguements.
And indeed they were passing.

That is not like I just write somethi.ng directly that is going directly to firebase.
I have checked it once running.

**Comment by lakshyajn** [comment]
> @lakshyajn first thing how your code is even running cause your package.json type is set to module but you are using require in index.js so just provide how you have tested your changes

Hey @sudoerp Thank U for that point, It just commented out😅
That should have happened when I was just importing and removing out other things for pushing.

@Satyam1923 sorry for that, the require functions should be just commented out, else everything works fine👍

**Comment by Satyam1923** [comment]
@lakshyajn Than being said your code still not running and what I observed you have started a branch with you started with you name. Kindly follow the guidelines provided for contributing in the project.
I'll also review the things which is leading to crash of your PR.

File: index.js [hunk 1]
```diff
@@ -1,11 +1,20 @@
 import express from "express";
 import axios from "axios";
 import bodyParser from "body-parser";
+import { registerUser, signInUser } from './public/js/firebase';    //Firebse Added
+import { registerUser, signInUser, signOutUser } from './public/js/auth'; //imported auth
```
Comment (Satyam1923) [inline index.js:5]: SyntaxError: Identifier 'registerUser' has already been declared

File: index.js [hunk 2]
```diff
@@ -84,7 +93,75 @@ app.get("/search1", async (req, res) => {
     });
   }
 });
+
+// So Firebase Authentication and Error Handling here->
+
+// Handle user registration
+const registerUser = (email, password) => {
```
Comment (Satyam1923) [inline index.js:100]: SyntaxError: Identifier 'registerUser' has already been declared

File: public/js/firebase.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import firebase from 'firebase/app';
+import 'firebase/auth';
+import 'firebase/firestore';
+
+const firebaseConfig = {
```
Comment (Satyam1923) [inline public/js/firebase.js:5]: Don't send your credentials in PRs. Do it with .env file

File: index.js [hunk 4]
```diff
@@ -1,11 +1,20 @@
 import express from "express";
 import axios from "axios";
 import bodyParser from "body-parser";
+import { registerUser, signInUser } from './public/js/firebase';    //Firebse Added
+import { registerUser, signInUser, signOutUser } from './public/js/auth'; //imported auth
+import { addData, getData, updateData, deleteData } from './public/js/database';  //imported database
+
+const { database, auth } = require('./public/js/firebase');
```
Comment (Satyam1923) [inline index.js:8]: Even if I comment it out reset of the code is still not working.

**Review by Satyam1923 — CHANGES_REQUESTED** [review]
Check all the comments I have listed out.

File: index.js (hunk 4 above)
Comment (lakshyajn) [reply index.js:8]: Ohh ok
Sorry for that
IDK why it ran in my machine
I'll just modify again👍

Comment (sudoerp) [reply index.js:8]: > Ohh ok Sorry for that IDK why it ran in my machine I'll just modify again👍

I think you are using a new Javascript runtime which we don't have access to ,cause Node won't be able to run this code

**Comment by lakshyajn** [comment]
Hey @Satyam1923, I did all the required changes 🙃
Have a look in it
U can add your firebase, as I just moved mine in .env file as u said.

Reminder what now have to be done in this code:-
give dummy values for arguments in registeruser, siginuser, signout user so that we can check
Also for addData, getData, updateData, deleteData, we need collection name, that is from where the data should be accessed or I should call it the path.

If u will run this code, u will get error handling in terminal, coz there is not data yet to show, like this->

Screenshot 2024-05-11 132358

As email, password are yet not defined + above arguments I mentioned above.

@sudoerp u may also check if u want and can suggest changes.
I can say I have completed it to the most.
If there are any changes left that would be minimal

And sorry for previous confusions,
in before commit, tried and tested was to mention that firebase was connected and can be used to pass data, I also tried to pass the data, but I have to make collectionName in realtime database to mention the path.

Screenshot 2024-05-11 131741

In that way along with collection, we can access data🤓

**Comment by Satyam1923** [comment]
@lakshyajn I cannot accept this because it will crash the main development for master branch. So what you can do is implement a working back-end Firebase environment .

**Pull request closed without merging**